File: windows/intel/interpreted/ruby/binary/binary_optimized.rb
def binary_search_optimized(arr, target, left = 0, right = nil)
  right = arr.length - 1 if right.nil?

  return -1 if left > right

  mid = (left + right) / 2
  if arr[mid] == target
    return mid
  elsif arr[mid] < target
    binary_search_optimized(arr, target, mid + 1, right)
  else
    binary_search_optimized(arr, target, left, mid - 1)
  end
end

# Test
arr = [2, 4, 6, 8, 10, 12, 14, 16, 18]
target = 14
index = binary_search_optimized(arr, target)
puts "Index: #{index}"

File: linux/intel/compiled/go/binary/binary_optimized.go
package main

import (
	"fmt"
	"time"
)

func binarySearchOptimized(arr []int, target int) int {
	low := 0
	high := len(arr) - 1

	for low <= high {
		// Avoid integer overflow:
		mid := low + (high-low)/2
		if arr[mid] == target {
			return mid
		} else if arr[mid] < target {
			low = mid + 1
		} else {
			high = mid - 1
		}
	}
	return -1
}

func main() {
	arr := []int{2, 4, 6, 8, 10, 12, 14, 16, 18}
	target := 14
	
	start := time.Now()
	index := binarySearchOptimized(arr, target)
	duration := time.Since(start)

	fmt.Printf("Target found at index: %d\n", index)
	fmt.Printf("Execution time: %v\n", duration)
}
